[CoreIPC] ASAN_TRAP in WebKit::ImageBufferShareableMappedIOSurfaceBackend::create

pratiksha-99 opened the pull request:
#### 45e0e3bd236dbae039329cfa14ce8a134e77f68f
<pre>
[CoreIPC] ASAN_TRAP in WebKit::ImageBufferShareableMappedIOSurfaceBackend::create
<a href="https://bugs.webkit.org/show_bug.cgi?id=286756">https://bugs.webkit.org/show_bug.cgi?id=286756</a>
<a href="https://rdar.apple.com/143264721">rdar://143264721</a>

Reviewed by Said Abou-Hallawa.

Added HAVE(IOSURFACE_RGB10) guards around all RGB10|A8 format usage.

* LayoutTests/ipc/create-image-buffer-crash-expected.txt: Added.
* LayoutTests/ipc/create-image-buffer-crash.html: Added.
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
(WebCore::convertToPixelFormat):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::PixelBuffer::supportedPixelFormat):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::makeVImageCGImageFormat):
(WebCore::convertImagePixelsSkia):
* Source/WebCore/platform/graphics/PixelFormat.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/PixelFormat.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
(WebKit::RemoteImageBufferSet::isOpaque const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::bytesPerPixel const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/289634@main">https://commits.webkit.org/289634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e12804f2f21b74dda639aa6e261e7a9456af53f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87557 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92420 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38301 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89608 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15240 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25378 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90559 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5697 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47981 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5483 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33632 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37414 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75889 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94308 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14725 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10804 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76464 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14979 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75088 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75687 "Found 2 new API test failures: /WebKitGTK/TestAuthentication:/webkit/Authentication/authentication-empty-realm, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18618 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20064 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18493 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7676 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14741 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20042 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14485 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17929 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pratiksha-99 commented on LayoutTests/ipc/create-image-buffer-crash-expected.txt:
```diff
@@ -0,0 +1,9 @@
+    RenderBody {BODY} at (8,8) size 784x576
+      RenderBlock {P} at (0,0) size 784x18
+        RenderText {#text} at (0,0) size 267x18
+          text run at (0,0) width 267: "This test passes if WebKit does not crash."
```
ignore this, fixing it.

rniwa commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -287,8 +287,17 @@ static RefPtr<ImageBuffer> allocateImageBufferInternal(const FloatSize& logicalS
-        if (!imageBuffer)
-            imageBuffer = ImageBuffer::create<ImageBufferShareableMappedIOSurfaceBackend, ImageBufferType>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext, imageBufferIdentifier);
+        if (pixelFormat == ImageBufferPixelFormat::RGB10 || pixelFormat == ImageBufferPixelFormat::RGB10A8) {
+#if HAVE(IOSURFACE_RGB10)
```
Is it correct that we fallback to unaccelerated IO surface when IOSURFACE_RGB10 is compiled out?

pratiksha-99 replied:
Not until `HAVE(IOSURFACE)` is also compiled out.

shallawa commented on LayoutTests/ipc/create-image-buffer-crash.html:
```diff
@@ -0,0 +1,42 @@
+      },
+      pixelFormat: 2,
+      renderingResourceIdentifier: 393230
+    });
```
Can we test all the `ImageBufferPixelFormat` values and make sure that none of them causes a crash?
```
function createRemoteImageBuffer(remoteBackend, format, identifier)
{
     ...
}

for (let format = 0; format < 5; ++format)
    createRemoteImageBuffer(remoteBackend, format, 393230 + format);
```

pratiksha-99 replied:
Updated the tests. Thanks.

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
```diff
@@ -80,7 +80,10 @@ class RemoteImageBufferSet : public IPC::StreamMessageReceiver, public ImageBuff
+#if HAVE(IOSURFACE_RGB10)
         return m_pixelFormat == WebCore::ImageBufferPixelFormat::RGB10 || m_pixelFormat == WebCore::ImageBufferPixelFormat::BGRX8;
+#endif
+        return m_pixelFormat == WebCore::ImageBufferPixelFormat::BGRX8;
```
I think two return statements in a function when `HAVE(IOSURFACE_RGB10)` is true is a little confusing. I think you either add `#else` and include the second return in it:

```
#if HAVE(IOSURFACE_RGB10)
        return m_pixelFormat == WebCore::ImageBufferPixelFormat::RGB10 || m_pixelFormat == WebCore::ImageBufferPixelFormat::BGRX8;
#else
        return m_pixelFormat == WebCore::ImageBufferPixelFormat::BGRX8;
#endif
```

Or remove the duplicate check:

```
#if HAVE(IOSURFACE_RGB10)
        if (m_pixelFormat == WebCore::ImageBufferPixelFormat::RGB10)
            return true;
#endif
        return m_pixelFormat == WebCore::ImageBufferPixelFormat::BGRX8;
```

I prefer the later.

pratiksha-99 committed:
[CoreIPC] ASAN_TRAP in WebKit::ImageBufferShareableMappedIOSurfaceBackend::create
https://bugs.webkit.org/show_bug.cgi?id=286756
rdar://143264721

Reviewed by Said Abou-Hallawa.

Added HAVE(IOSURFACE_RGB10) guards around all RGB10|A8 format usage.

* LayoutTests/ipc/create-image-buffer-crash-expected.txt: Added.
* LayoutTests/ipc/create-image-buffer-crash.html: Added.
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
(WebCore::convertToPixelFormat):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::PixelBuffer::supportedPixelFormat):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::makeVImageCGImageFormat):
(WebCore::convertImagePixelsSkia):
* Source/WebCore/platform/graphics/PixelFormat.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/PixelFormat.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
(WebKit::RemoteImageBufferSet::isOpaque const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::bytesPerPixel const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/289634@main
